pySOT/utils.py

```diff
@@ -390,3 +393,267 @@ def optimize(self):
+    :type F: 2D numpy array
+    :return: 1) ndf_index --> indices of nd points in F,
+             2) df_index --> indices of dominated points in F
+    :rtype: lists of integers
+    """
+
+    (M, N) = F.shape
+    df_index = []
+    ndf_index = [int(0)]
+    for i in range(1, N):
+        (ndf_index, df_index) = ND_Add(F[:, 0:i+1], df_index, ndf_index)
+    return (ndf_index, df_index)
+
+
+def nd_sorting(F, nmax):
+    """ Computes the non-domination rank of each row vector in F
+
+    This function sorts the objective functions vectors according to
+    non-domination rank. If F is too large, sorting is only done for
+    the first nmax points
+
+    :param F: 2D array (MxN) containing a set of N vectors of M objectives
+    :type F: numpy array
+    :param nmax: maximum number of points ranked via nd_sorting
+    :type nmax: int
+    :return: list of N integers indicating rank of each row vector in F
+    :rtype: list of size N
+    """
+
+    (M, N) = F.shape
+    nd_ranks = np.ones((N, ), dtype=np.int)*POSITIVE_INFINITY
+    P = np.ones((N, ), dtype=np.int)
+    for i in range(0, N):
+        P[i] = i
+    i = 1
+    count = 0
+    while count < nmax and len(P) > 0:
+        (ndf_index, df_index) = ND_Front(F[:, P])
+        for j in range(0, len(ndf_index)):
+            nd_ranks[P[ndf_index[j]]] = i
+        count = count + len(ndf_index)
+        P_new = np.ones((len(df_index), ), dtype=np.int)
+        for j in range(0, len(df_index)):
+            P_new[j] = P[df_index[j]]
+        P = P_new
+        i = i+1
+    return nd_ranks
+
+
+def check_radius_rule(X, X_c, sigma, dim, nc, d_thresh=1.0):
+    """ Function that implements the Radius Rule of SOP
+
+    This function implements the radius-rule (MODIFIED) for SOP algorithm
+    by checking if a new potential center is within a threshold radius of
+    previously selected centers or not.
+
+    :param X: A vector/point being considered to be selected as a center
+    :type X: numpy 1D array
+    :param X_c: Set of already selected centers
+    :type X_c: 2D numpy array
+    :param sigma: Candidate search radius
+    :type sigma: Float
+    :param dim: number of decision variables in X
+    :type dim: int
+    :param nc: number of centers (i.e., rows) in X_c
+    :type nc: int
+    :param d_thresh: An adaptive multiplier that allows points close to each
+        other to be selected as center points, simultaneously
+    :type d_thresh: float
+    :return: A flag with value = 0 if X is within a distance sigma*d_thresh
+        of any point in X_c and 1 otherwise
+    :rtype: Bool
+    """
+
+    flag = 1
+    for i in range(nc):
+        if X_c[i, dim+4] == 0:
+            sigma_new = sigma
+        else:
+            sigma_new = np.power(1/2, X_c[i, dim+4])*sigma
+        d = scp.distance.euclidean(X, X_c[i, 0:dim])
+        if d < sigma_new*d_thresh/np.sqrt(len(X)):
+            flag = 0
+            break
+    return flag
+
+
+class SopRecord():
```

Comment:
Can you move this back to `strategy.py` and rename it `_SopRecord`? Same for `SopCenter`

Reply:
Done